File: neural-house/apps/api/app/schemas/audition.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field


class AuditionProviderConfig(BaseModel):
    provider: str = "local_stub"
    api_base_url: str | None = None
    model_name: str = "deterministic-fallback"
    enabled: bool = False


class AuditionSkinConfig(BaseModel):
    palette: str = "studio_blue"
    accent: str = "gold"
    silhouette: str = "host-ready"
    hair_style: str = "crown"


class AuditionAgentConfig(BaseModel):
    character_name: str
    archetype: str
    speech_style: str
    public_hook: str
    traits: list[str] = Field(default_factory=list)
    strengths: list[str] = Field(default_factory=list)
    weaknesses: list[str] = Field(default_factory=list)
    skin: AuditionSkinConfig = Field(default_factory=AuditionSkinConfig)


class AuditionCreateRequest(BaseModel):
    provider_config: AuditionProviderConfig = Field(default_factory=AuditionProviderConfig)
    agent_config: AuditionAgentConfig
    simulated_minutes: int = Field(default=6, ge=2, le=12)
    playback_ms_per_beat: int = Field(default=5000, ge=500, le=15000)


class AuditionEventRead(BaseModel):
    id: int
    session_id: int
    tick_number: int
    simulated_second: int
    zone: str
    action_type: str
    summary: str
    dialogue: str | None
    state_snapshot_json: dict
    created_at: datetime


class AuditionSessionRead(BaseModel):
    id: int
    status: str
    environment_code: str
    environment_label: str
    provider: str
    api_base_url: str | None
    model_name: str
    provider_enabled: bool
    character_name: str
    archetype: str
    speech_style: str
    public_hook: str
    traits_json: list[str]
    strengths_json: list[str]
    weaknesses_json: list[str]
    skin_palette: str
    skin_accent: str
    skin_silhouette: str
    hair_style: str
    simulated_minutes: int
    playback_ms_per_beat: int
    total_beats: int
    current_beat: int
    summary: str
    session_seed: int
    created_at: datetime
    started_at: datetime
    ended_at: datetime | None


class AuditionLiveState(BaseModel):
    session: AuditionSessionRead
    visible_events: list[AuditionEventRead]
    next_event_eta_ms: int | None

File: neural-house/packages/contracts/src/audition.ts
import { z } from "zod";

import { safeNumberSchema, strictObject, summarySchema, titleSchema, urlSchema } from "./common.js";
import { providerNameSchema } from "./provider.js";

export const auditionStatusSchema = z.enum(["draft", "queued", "running", "completed", "failed"]);
export const auditionZoneSchema = z.enum([
  "entrance",
  "host_mark",
  "camera_lane",
  "spotlight",
  "confessional_cam",
  "exit_mark"
]);

export const auditionAgentSkinSchema = strictObject({
  palette: z.string().trim().min(1).max(40),
  accent: z.string().trim().min(1).max(40),
  silhouette: z.enum(["host-ready", "sharp", "soft", "masked", "android"]),
});

export const auditionProviderConfigSchema = strictObject({
  provider: providerNameSchema,
  apiBaseUrl: urlSchema.optional().or(z.literal("")).transform((value) => value || undefined),
  modelName: titleSchema,
  enabled: z.boolean().default(false),
});

export const auditionAgentConfigSchema = strictObject({
  characterName: titleSchema,
  archetype: z.string().trim().min(1).max(80),
  speechStyle: summarySchema,
  publicHook: summarySchema,
  traits: z.array(z.string().trim().min(1).max(60)).min(1).max(8),
  strengths: z.array(z.string().trim().min(1).max(80)).default([]),
  weaknesses: z.array(z.string().trim().min(1).max(80)).default([]),
  skin: auditionAgentSkinSchema,
});

export const auditionCreateRequestSchema = strictObject({
  providerConfig: auditionProviderConfigSchema,
  agentConfig: auditionAgentConfigSchema,
  simulatedMinutes: safeNumberSchema.int().min(2).max(12).default(6),
  playbackMsPerBeat: safeNumberSchema.int().min(500).max(15000).default(5000),
});

export const auditionEventSchema = strictObject({
  id: safeNumberSchema.int().positive(),
  sessionId: safeNumberSchema.int().positive(),
  tickNumber: safeNumberSchema.int().nonnegative(),
  simulatedSecond: safeNumberSchema.int().nonnegative(),
  zone: auditionZoneSchema,
  actionType: z.string().trim().min(1).max(80),
  summary: summarySchema,
  dialogue: summarySchema.optional(),
  stateSnapshot: strictObject({
    confidence: safeNumberSchema.min(0).max(100),
    stress: safeNumberSchema.min(0).max(100),
    cameraHeat: safeNumberSchema.min(0).max(100),
  }),
});

export const auditionSessionSchema = strictObject({
  id: safeNumberSchema.int().positive(),
  status: auditionStatusSchema,
  environmentLabel: z.string().trim().min(1).max(120),
  providerConfig: auditionProviderConfigSchema,
  agentConfig: auditionAgentConfigSchema,
  simulatedMinutes: safeNumberSchema.int().positive(),
  playbackMsPerBeat: safeNumberSchema.int().positive(),
  totalBeats: safeNumberSchema.int().positive(),
  currentBeat: safeNumberSchema.int().nonnegative(),
  summary: summarySchema,
  createdAt: z.string().datetime({ offset: true }),
  startedAt: z.string().datetime({ offset: true }).optional(),
  endedAt: z.string().datetime({ offset: true }).optional(),
});

export type AuditionStatus = z.infer<typeof auditionStatusSchema>;
export type AuditionZone = z.infer<typeof auditionZoneSchema>;
export type AuditionProviderConfig = z.infer<typeof auditionProviderConfigSchema>;
export type AuditionAgentConfig = z.infer<typeof auditionAgentConfigSchema>;
export type AuditionCreateRequest = z.infer<typeof auditionCreateRequestSchema>;
export type AuditionEvent = z.infer<typeof auditionEventSchema>;
export type AuditionSession = z.infer<typeof auditionSessionSchema>;
